### PR TITLE
DRA: allow Cluster Autoscaler to integrate with DRA scheduler plugin

### PR DESCRIPTION
**Pull request by towca**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR refactors the DRA scheduler plugin so that all accesses to DRA objects, and all in-memory tracking for ResourceClaim modifications, go through a `SharedDRAManager` interface that can be plugged into the scheduler framework. 

This is needed for Cluster Autoscaler to integrate with DRA. Cluster Autoscaler will plug a different `SharedDRAManager` implementation.

Plugging `SharedDRAManager` allows CA to make modifications to the DRA objects provided to the DRA plugin:
* Injecting fake node-local `ResourceSlices`  - when a Node exposing `ResourceSlices` is duplicated in-memory (e.g. during a scale-up simulation).
* Removing node-local `ResourceSlices` - when a Node exposing `ResourceSlices` is deleted in-memory (e.g. during a scale-down simulation).
* Injecting fake pod-owned `ResourceClaims` - when a Pod owning `ResourceClaims` is duplicated in-memory (e.g. a DS pod during a scale-up simulation).
* Removing pod-owned `ResourceClaims` - when a Pod owning `ResourceClaims` is deleted in-memory (e.g. for the concept of "expendable pods").
* Modifying `ResourceClaims` - e.g. clearing the allocation, or modifying `ReservedFor`, when simulating where else a given scheduled pod can be "rescheduled" .

Plugging `SharedDRAManager` also allows CA to obtain ResourceClaim allocations computed during `Filter` back from the DRA plugin.

#### Which issue(s) this PR fixes:

The CA/DRA integration is tracked in #118612. The integration requires changes in CA and kube-scheduler - this is the scheduler part. The corresponding CA changes are in https://github.com/kubernetes/autoscaler/pull/7350.

#### Special notes for your reviewer:

1. Some parts of `SharedDRAManager` aren't strictly necessary for CA integration - e.g. CA will never modify `DeviceClasses`, so that could technically stay as just a lister. Having all DRA objects in the interface simplifies testing on the CA side. Instead of having to fiddle with faking informers separately for `DeviceClasses`, they can just be injected easily through `SharedDRAManager`, like claims and slices. IMO it also makes for cleaner code overall.
2. Some of `ResourceClaimTracker` methods (`RemoveClaimPendingAllocation`, `AssumeClaimAfterApiCall`, `AssumedClaimRestore`) are only applicable to the binding cycle, so they'll never be run by Cluster Autoscaler (which only runs the scheduling cycle phases). It'd be nice to separate them into 2 interfaces or do something similar, instead of just not implementing a subset of the methods in CA. But that's not great from readability perspective, as we'd have tightly coupled methods in separate interfaces. I'd appreciate any ideas on how to separate the methods for scheduling/binding phases while maintaining readability.
3. The added layer of abstraction in the DRA plugin decreases readability a bit (`ResourceClaimTracker` coupling unrelated mechanisms together is not ideal..), but it was the cleanest solution I found while prototyping. CA definitely needs to be able to plug into the listers, and to have some way of getting an allocation result back. I didn't find a good solution that satisfies the 2 properties without extracting the cache + the in-flight map + the listers behind a single interface.
4. I know this conflicts with #127277. From a quick read, it should be fairly straightforward to rebase one on top of the other. I can rebase after #127277 is submitted, just wanted to get this PR out together with the CA PR. Also note that I'm OOO from Oct 11th to Oct 27th, so I won't have a lot of time to iterate on it before the 1.32 code freeze date.

#### Does this PR introduce a user-facing change?

Not on its own, the PR just makes Cluster Autoscaler integration possible.

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/9de7f62e16fc5c1ea3bd40689487c9edc7fa5057/keps/sig-node/4381-dra-structured-parameters/README.md
```

### CODE REVIEW

File: pkg/scheduler/framework/interface.go
Comment (nojnhuh) [inline pkg/scheduler/framework/interface.go:826]: Since this method was added in 1.31 by #124595, are we sure we can remove this in 1.32?

File: pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go
Comment (jackfrancis) [inline pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go:1]: we need this stuff here

```
/*
Copyright 2024 The Kubernetes Authors.
...
```

File: pkg/scheduler/framework/listers.go
Comment (jackfrancis) [inline pkg/scheduler/framework/listers.go:100]: Is `found` the right descriptor for this bool? Would `deleted` be better?

File: pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go
Comment (jackfrancis) [inline pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:1289]: ditto my comment above, now that we have a single bool return value, should we rename `found` to something that better describes the behavior, like `deleted`?

File: pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go
Comment (jackfrancis) [inline pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go:132]: Are we anticipating another `ResourceClaimTracker` implementation might benefit from an error response? Curious why the introduction of the `err` return value wrapping around the `Map.Store` behavior, which doesn't return a value.

**Comment by towca** [comment]
@kubernetes/sig-scheduling-leads I'd also appreciate if you could take a look here. In particular, is it ok to just replace `AssumeCache` with `SharedDraManager` in the framework this way? Or do we have to follow some deprecation process?

File: pkg/scheduler/framework/interface.go
Comment (towca) [reply pkg/scheduler/framework/interface.go:826]: Good point, thanks for bringing this up.

I assumed so since it was just added specifically for the DRA plugin, if we leave it won't be used by any in-tree plugins. I guess in theory we could have out-of-tree plugins that started to depend on it right as it was introduced?

@kubernetes/sig-scheduling-leads Could you provide guidance here?

**Comment by alculquicondor** [comment]
cc @dom4ha

File: pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go
Comment (towca) [reply pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go:148]: Yeah, the CA implementation returns an error from this if it isn't tracking the claim. This should never happen normally, but it was helpful when debugging the initial implementation.

The error isn't necessary in the scheduler implementation since we're not creating any fake ResourceClaims in that case.

Comment (jackfrancis) [reply pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go:148]: ACK

Comment (MaciekPytel) [reply pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go:148]: Should we mention this in a comment? I can foresee someone cleaning up an unused return value without looking at CA code (likely without realizing they should even look at autoscaler repo)

File: pkg/scheduler/framework/listers.go
Comment (MaciekPytel) [inline pkg/scheduler/framework/listers.go:76]: I'm not sure to what extent those comments are helpful. It feels like instead of describing the expected behavior they describe a reference implementation and rely on the user / implementer to be familiar enough with all the underlying concepts to infer the behavior of said implementation. I'd consider rephrasing the comments here to describe the resulting behavior that a method user should expect and rely on people implementing this interface to figure out how to achieve expected behavior - they can find reference implementation easily enough if they want to.

For example - I'd replace `If a real informer is used, changes made via AssumeClaimAfterAPICall() should be applied to the obtained claims until they're reflected in the informer cache.` with something along the lines of `The result is guaranteed to immediately include any changes made via AssumeClaimAfterAPICall() in order to avoid potential races caused by communication with api-server`.

Comment (MaciekPytel) [reply pkg/scheduler/framework/listers.go:76]: I'd also rephrase other comments in ResourceClaimTracker interface definition in a similar way

**Comment by MaciekPytel** [comment]
I left some minor comments, but overall this lgtm from the perspective of integration with Cluster Autoscaler / sig-autoscaling. I'll leave applying actual lgtm label to owners.

**Comment by sftim** [comment]
This change would be visible to end users: they could observe the different between a cluster with and without the change that this PR makes.

Hence, it needs a changelog entry.

**Comment by alculquicondor** [comment]
Can you elaborate? Which kind of observable difference will there be?

**Comment by jackfrancis** [comment]
@alculquicondor I think the title of the PR would be sufficient to describe the difference between

before:
- I'm using DRA / ResourceClaims on nodes that are in a cluster-autoscaler-enabled pool, but there is clearly not graceful interaction between the two
after:
- My ResourceClaim resource lifecycles are initiating sane autoscaling behaviors using the cluster-autoscaler

The only gotcha here is that these behavioral changes are only gonna manifest when you're using a version of cluster-autoscaler that has DRA integration in it (https://github.com/kubernetes/autoscaler/pull/7350), which we hope to land in a v1.32 release of cluster-autoscaler. But this k/k PR wants to land in v1.32 as well.

So if you're using a v1.32 of cluster-autoscaler w/ the DRA integrations on a v1.31 k8s cluster, then the user will observe changes when upgrading their k8s cluster to v1.32. But you should not ever run a n+1 version of cluster-autoscaler on a n version of k8s, so it might be academic.

Hope that helps.

**Comment by alculquicondor** [comment]
Right, but those changes are fully dependent on what cluster-autoscaler version the user is running. The PR in this repo doesn't seem to enable anything from an end-user perspective. Thus, a release note NONE is adequate.

**Comment by sftim** [comment]
> So if you're using a v1.32 of cluster-autoscaler w/ the DRA integrations on a v1.31 k8s cluster, then the user will observe changes when upgrading their k8s cluster to v1.32. 

Those are the changes we should changelog. Even if a user who follows our guidance wouldn't see any problems, the control plane's behavior has changed [will have changed] between minor releases.

**Comment by alculquicondor** [comment]
> So if you're using a v1.32 of cluster-autoscaler w/ the DRA integrations on a v1.31 k8s cluster, then the user will observe changes when upgrading their k8s cluster to v1.32.

This statement is not accurate. There shouldn't be a behavior change when upgrading the k8s control plane related to this change.

This just allows cluster-autoscaler to implement certain logic, but that's still to be implemented in cluster-autoscaler. We are not talking about an API that communicates kube-scheduler and cluster-autoscaler. This is a framework change that cluster-autoscaler can utilize in code.

**Comment by pohly** [comment]
> Do we changelog framework changes?

There's no consensus on whether Go API changes are worth a changelog entry. I would say it's judgment call that depends on whether it silently breaks someone (downstream consumers *must* update their code because it compiles, but doesn't work as intended - we should never allow such changes!), who is impacted (more important in client-go, not so much here), and how obvious the resolution is. Reporting Go API changes is just noise for most of our downstream consumers, so I would set a pretty high bar before including it.

TLDR: I wouldn't not include this change.

**Comment by jackfrancis** [comment]
Based on OWNERS data, it looks like we need @klueska @pohly or @bart0sh to weigh in for sig-node, @MaciekPytel or @x13n for sig-autoscaling and @alculquicondor can represent sig-scheduling. We wanna land this in v1.32 cycle, let us know if this needs any additional work, thanks!

**Comment by alculquicondor** [comment]
@dom4ha can you give a first pass?

File: pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go
Comment (pohly) [inline pkg/scheduler/framework/plugins/dynamicresources/dra_manager.go:32]: Nit: I probably would call this `SharedDRAManager`.

Linters complain when `URL` is spelled `Url` in a variable or method name. The same applies here.

File: staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go
Comment (pohly) [inline staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:44]: In https://github.com/kubernetes/kubernetes/pull/127277/files#diff-8907e59dbe7c433690f8025b6cc248893a60478f6e751d4679bf0e64aa4471e7, I am completely removing this "list all slices" step. Instead, the caller needs to track allocated devices, which turned out to be more efficient because tracking can be incremental instead of always starting from scratch.

I'm on the edge whether I should ask you to hold this PR until mine is reviewed and merged, or whether I should let you go ahead and then update mine. But I would need to change the interfaces that you are implementing in CA, so perhaps it is better to get my changes in first so that you don't need to rewrite the CA code?

https://github.com/kubernetes/kubernetes/pull/127277 is blocked by https://github.com/kubernetes/kubernetes/pull/127266. Perhaps I should rebase #127277 on top of master so that we can merge it sooner? I'd prefer not to, but it would be possible.

Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:44]: I forgot to mention that apart from these logistics the proposed changes look okay to me.

Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:44]: I rebased https://github.com/kubernetes/kubernetes/pull/127277. Perhaps someone from SIG Scheduling can review the new commits in that PR (i.e. ignore the DRA API changes)? Then when those API changes are merged, #127277 could be merged quickly (rebase, re-add LGTM).

Comment (jackfrancis) [reply staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:44]: @pohly I would suggest you land your PR changes first and we can adapt this one on top of it. @towca is on vacation for another week or so, so you can take the lead position.

If https://github.com/kubernetes/kubernetes/pull/127266 and https://github.com/kubernetes/kubernetes/pull/127277 land early next week, then SIG Autoscaling could spawn a new PR for this CA integration from this one, incorporating your feedback and the changes necessitated by your new DRA foo in master.

cc @MaciekPytel @nojnhuh @ritazh

Comment (pohly) [reply staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:44]: #127266 might land early next week, but for #127277 I still need a reviewer from SIG scheduling for the final LGTM and approval. It was reviewed already by a member of WG Device management. Any volunteers?

Comment (jackfrancis) [reply staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:44]: cc @alculquicondor @Huang-Wei @ahg-g @kerthcet @sanposhiho

File: test/integration/scheduler_perf/dra.go
Comment (jackfrancis) [reply test/integration/scheduler_perf/dra.go:28]: unfortunate that gofmt doesn't seem to care about this :/

Comment (pohly) [reply test/integration/scheduler_perf/dra.go:28]: `gofmt` or some variant of it sorts imports. But splitting up imports into related blocks is left to the developer. Typically we use
```
<standard library>

<third party packages>

<k8s.io packages>
```

File: pkg/scheduler/framework/interface.go
Comment (towca) [reply pkg/scheduler/framework/interface.go:826]: @dom4ha @alculquicondor Any idea here? Is this ok?

File: staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go
Comment (towca) [reply staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:44]: Hmm, did you mean removing the "list all allocated claims" step? From a quick read of #127277, the allocator still needs all ResourceSlices + all allocated devices from allocated ResourceClaims, right?

So `framework.ResourceClaimLister` can stay as proposed, but in `framework.ResourceClaimTracker` we'd need to replace `ListAllAllocated()` with something like `ListAllAllocatedDevices()`? If so, that's a trivial change in CA code - the first implementation can just create the list from scratch every time. So I'm okay with merging in either order, but we don't have much time.

**Comment by towca** [comment]
Addressed all comments! IIUC, the current state is:
* We need another review pass and approval from sig-scheduling @dom4ha @alculquicondor
  * In particular, it's not clear whether we can just remove the just-added AssumeCache from the framework.
* We need another review pass and approval from sig-autoscaling @MaciekPytel
* We need to decide whether to merge this or #127277 first @pohly

**Comment by MaciekPytel** [comment]
> We need another review pass and approval from sig-autoscaling @MaciekPytel

lgtm. I won't use the bot command form, as sig-scheduling is the owner here (also it would be removed when you rebase anyway)

File: staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go
Comment (dom4ha) [reply staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go:44]: Yes, list all allocated claims `ListAllAllocated()` need to be replaced by `ListAllAllocatedDevices()`, but it will be still backed by assume cache, so I believe the interface can remain the same.

The change that #127277 brings is that the device list (previously claim list) is snapshotted in PreFilter and used in Filter as static data, so the device lister (previously claim lister) does not need to be passed here. The same applies to resource slice lister. The class lister will be the only remaining one that needs to be passed.

**Review by dom4ha — COMMENTED** [review]
The changes looks good from the scheduler perspective, so not leaving lgtm yet.
I think it would be easier if #127277 goes first, as it would be more complex to rebase there.

File: pkg/scheduler/framework/interface.go
Comment (towca) [reply pkg/scheduler/framework/interface.go:826]: Ack, thanks for the confirmation!

File: staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go
Comment (towca) [reply staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go]: Done

**Comment by towca** [comment]
#127277 was submitted, I rebased this PR.

@pohly @dom4ha @alculquicondor Could you take another look?

**Review by pohly — COMMENTED** [review]
This looks good to me. Thanks for your patience and support!

**Review by jackfrancis — APPROVED** [review]
my comments are addressed, thank you @towca!

@alculquicondor @dom4ha we'll need an actual lgtm from scheduing

@pohly I think you can lgtm for sig-node

**Comment by pohly** [comment]
/assign @alculquicondor @Huang-Wei 

For approval.

**Review by dom4ha — COMMENTED** [review]
/lgtm

**Review by kerthcet — COMMENTED** [review]
Except the interfaces, most of the modifications are **Replace** OPs.
/lgtm
/approve

Thanks!

**Pull request merged**
